# Project 03 -- Ansh Tandon

**TA Help:** John Smith, Alice Jones

- Help with figuring out how to write a function.
    
**Collaboration:** Friend1, Friend2
    
- Helped figuring out how to load the dataset.
- Helped debug error with my plot.

## Question 1

In [3]:
import pandas as pd

In [4]:
from pathlib import Path

In [5]:
Path("/anvil/projects/tdm/data/noaa/1750.csv")

PosixPath('/anvil/projects/tdm/data/noaa/1750.csv')

In [6]:
import os

In [9]:
len(os.listdir('/anvil/projects/tdm/data/noaa/'))

266

In [10]:
myfiles=[]
for year in range (1880, 1884):
    file= Path(f'/anvil/projects/tdm/data/noaa/{year}.csv')
    myfiles.append(file)
print(myfiles)

[PosixPath('/anvil/projects/tdm/data/noaa/1880.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1881.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1882.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1883.csv')]


In [22]:
# the below uses list comprehension
myfiles=[Path(f'/anvil/projects/tdm/data/noaa/{year}.csv') for year in range (1880, 1884)]

In [23]:
print(myfiles)

[PosixPath('/anvil/projects/tdm/data/noaa/1880.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1881.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1882.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1883.csv')]


1) For the files, there is 1 file for 1750 and 261 files from 1763 to 2023 (excluding 2020 and .txt files). As there is also a 2020_sample.csv, 2020_sampleB.csv, readme-by-year.txt, and status-by-year.txt, that would bring the total files that arein the directory to 266 files. These files contain information about various records in the given year.

## Question 2

In [24]:
# the below will calculate the amount of records for 1880
file = "/anvil/projects/tdm/data/noaa/1880.csv"
with open(file,"r") as f:
    mycount = 0
    for line in f:
        mycount += 1
print(f'There are {mycount} records in the file called {file}')

There are 370779 records in the file called /anvil/projects/tdm/data/noaa/1880.csv


In [27]:
total_count = 0
for file in myfiles:
    with open(file,"r") as f:
        mycount = 0
        for line in f:
            total_count += 1
            mycount += 1
    print(f'There are {mycount} records in the file called {file}\n')

There are 370779 records in the file called /anvil/projects/tdm/data/noaa/1880.csv

There are 443270 records in the file called /anvil/projects/tdm/data/noaa/1881.csv

There are 514083 records in the file called /anvil/projects/tdm/data/noaa/1882.csv

There are 578730 records in the file called /anvil/projects/tdm/data/noaa/1883.csv



In [30]:
print(f'There are {total_count} records in the 4 files altogether.\n')

There are 1906862 records in the 4 files altogether.



Markdown notes and sentences and analysis written here.

## Question 3

In [32]:
myDF = pd.read_csv(myfiles[0])

In [34]:
myDF.columns

Index(['AGE00147705', '18800101', 'PRCP', '0', 'Unnamed: 4', 'Unnamed: 5', 'E',
       'Unnamed: 7'],
      dtype='object')

In [38]:
myDF.head()

AGE00147705  18800101  PRCP   0 Unnamed: 4 Unnamed: 5  E  Unnamed: 7
0  AGE00147708  18800101  PRCP  36        NaN        NaN  E         NaN
1  AGE00147709  18800101  PRCP  36        NaN        NaN  E         NaN
2  AGE00147712  18800101  PRCP   0        NaN        NaN  E         NaN
3  AGE00147713  18800101  TMAX  52        NaN        NaN  E         NaN
4  AGE00147713  18800101  TMIN   0        NaN        NaN  E         NaN

In [39]:
myDF.tail()

AGE00147705  18800101  PRCP    0 Unnamed: 4 Unnamed: 5  E  Unnamed: 7
370773  USW00093852  18801231  TMAX   39        NaN        NaN  X         NaN
370774  USW00093852  18801231  TMIN  -17        NaN        NaN  X         NaN
370775  USW00094728  18801231  TMAX -122        NaN        NaN  0         NaN
370776  USW00094728  18801231  TMIN -194        NaN        NaN  0         NaN
370777  USW00094728  18801231  PRCP    0        NaN        NaN  0         NaN

In [40]:
myDF = pd.read_csv(myfiles[0], header=None)

In [41]:
myDF.columns

Index([0, 1, 2, 3, 4, 5, 6, 7], dtype='int64')

In [42]:
myDF.head()

0         1     2   3    4    5  6   7
0  AGE00147705  18800101  PRCP   0  NaN  NaN  E NaN
1  AGE00147708  18800101  PRCP  36  NaN  NaN  E NaN
2  AGE00147709  18800101  PRCP  36  NaN  NaN  E NaN
3  AGE00147712  18800101  PRCP   0  NaN  NaN  E NaN
4  AGE00147713  18800101  TMAX  52  NaN  NaN  E NaN

In [43]:
myDF.tail()

0         1     2    3    4    5  6   7
370774  USW00093852  18801231  TMAX   39  NaN  NaN  X NaN
370775  USW00093852  18801231  TMIN  -17  NaN  NaN  X NaN
370776  USW00094728  18801231  TMAX -122  NaN  NaN  0 NaN
370777  USW00094728  18801231  TMIN -194  NaN  NaN  0 NaN
370778  USW00094728  18801231  PRCP    0  NaN  NaN  0 NaN

In [44]:
myfiles=[]
for year in range (1880, 1884):
    file= Path(f'/anvil/projects/tdm/data/noaa/{year}.csv')
    myfiles.append(file)
print(myfiles)

[PosixPath('/anvil/projects/tdm/data/noaa/1880.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1881.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1882.csv'), PosixPath('/anvil/projects/tdm/data/noaa/1883.csv')]


In [45]:
myfiles[0]

PosixPath('/anvil/projects/tdm/data/noaa/1880.csv')

In [51]:
newDF = pd.read_csv(myfiles[0],names = ["id","date","element_code","value","mflag","qflag","sflag","obstime"])

In [52]:
newDF

id      date element_code  value mflag qflag sflag  obstime
0       AGE00147705  18800101         PRCP      0   NaN   NaN     E      NaN
1       AGE00147708  18800101         PRCP     36   NaN   NaN     E      NaN
2       AGE00147709  18800101         PRCP     36   NaN   NaN     E      NaN
3       AGE00147712  18800101         PRCP      0   NaN   NaN     E      NaN
4       AGE00147713  18800101         TMAX     52   NaN   NaN     E      NaN
...             ...       ...          ...    ...   ...   ...   ...      ...
370774  USW00093852  18801231         TMAX     39   NaN   NaN     X      NaN
370775  USW00093852  18801231         TMIN    -17   NaN   NaN     X      NaN
370776  USW00094728  18801231         TMAX   -122   NaN   NaN     0      NaN
370777  USW00094728  18801231         TMIN   -194   NaN   NaN     0      NaN
370778  USW00094728  18801231         PRCP      0   NaN   NaN     0      NaN

[370779 rows x 8 columns]

In [53]:
mydataframes = []
for i in range(0,4):
    mydataframes.append(pd.read_csv(myfiles[i],names = ["id","date","element_code","value","mflag","qflag","sflag","obstime"]))

In [66]:
for i in range(0,4):
    print(mydataframes[i].shape)

(370779, 8)
(443270, 8)
(514083, 8)
(578730, 8)


In [67]:
for i in range(0,4):
    print(mydataframes[i].columns)

Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')
Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')
Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')
Index(['id', 'date', 'element_code', 'value', 'mflag', 'qflag', 'sflag',
       'obstime'],
      dtype='object')


1) Yes I do see something unexpected. The columns names from doing the myDF.columns, head(), and tail() function are talking one of the actual row values as a names of the columns, rather than actually indexing the dataframe to be correct similar to how the rows are ordered.

## Question 4

In [68]:
len(mydataframes)

4

In [73]:
for i in range(0,4):
    print(mydataframes[i]['element_code'].unique())
    print("\n")

['PRCP' 'TMAX' 'TMIN' 'SNOW' 'SNWD' 'TAVG' 'MDPR' 'DATN' 'DATX' 'MDTN'
 'MDTX' 'DAPR' 'DWPR' 'MDSF' 'WT01' 'TOBS' 'WT04' 'WT05']


['PRCP' 'TMAX' 'TMIN' 'TAVG' 'SNOW' 'SNWD' 'MDPR' 'MDSF' 'DATN' 'DATX'
 'MDTN' 'MDTX' 'WT04' 'DAPR' 'DWPR' 'TOBS']


['PRCP' 'TMAX' 'TMIN' 'TAVG' 'SNOW' 'SNWD' 'DATN' 'DATX' 'MDTN' 'MDTX'
 'DAPR' 'DWPR' 'MDPR' 'MDSF' 'WT04' 'TOBS' 'WT05' 'WT03' 'WT16' 'WT18'
 'DASF']


['PRCP' 'TMAX' 'TMIN' 'TAVG' 'SNOW' 'SNWD' 'DAPR' 'DWPR' 'MDPR' 'MDSF'
 'WT11' 'DATN' 'DATX' 'MDTN' 'MDTX' 'WT03' 'DASF' 'WT18' 'WT04' 'WT16'
 'WT08']




In [77]:
# Number of unique elements in each of the four dadtaframes
for i in range(0,4):
    print(len(mydataframes[i]['element_code'].unique()))

18
16
21
21


In [78]:
# Number of times SNOW occurs in each of the four dataframes
for i in range(0,4):
    print(mydataframes[i]['element_code'].value_counts()['SNOW'])

33700
36451
40648
53417


Markdown notes and sentences and analysis written here.

## Question 5

In [79]:
import pandas as pd
from pathlib import Path
myfiles=[]
for year in range (1880, 1884):
    file= Path(f'/anvil/projects/tdm/data/noaa/{year}.csv')
    myfiles.append(file)

In [80]:
count = 0
for file in myfiles:
    for myDF in pd.read_csv(file,names=["id","date","element_code","value","mflag","qflag","sflag","obstime"],chunksize =10000):
        count += len(myDF[myDF['element_code'] == 'SNOW'])

print(count)

164216


In [81]:
count = 0
for file in myfiles:
    for myDF in pd.read_csv(file,names=["id","date","element_code","value","mflag","qflag","sflag","obstime"],chunksize =10000):
        for index, row in myDF.iterrows():
            if row['element_code'] == 'SNOW':
                count += 1

print(count)

164216


Chunks size basically allows you to read a a small amount of data at a time so that you do not read everything in a given file at once, which would be easier when processing data with over million rows of data.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.